### PR TITLE
Fix gomobile toolchain bug

### PR DESCRIPTION
**Pull request by corfe83**
This PR solves:

#3128

The reason this happens is that runtime.Version() returns this value when loopvar variable is set:

"go1.23.2 X:loopvar"

### CODE REVIEW

**Comment by hajimehoshi** [comment]
Interesting.. Thanks!

File: cmd/ebitenmobile/gomobile.go [hunk 1]
```diff
@@ -164,7 +165,8 @@ import (
 	// runtime.Version() is the current executing Go version. For example, this is the version of the toolchain directive in go.mod.
 	// This might differ from the Go command version under the temporary directory.
 	// To avoid the version mismatch, set the toolchain explicitly (#3086).
-	if err := runGo("mod", "edit", "-toolchain="+runtime.Version()); err != nil {
+	toolchainVersion := strings.Split(runtime.Version(), " ")[0]
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:168]:
```go
toolchainVersion, _, _ := strings.Cut(runtime.Version(), " ")
```

is better.

Comment (hajimehoshi) [reply cmd/ebitenmobile/gomobile.go:168]: Optional: a more robust way would be

```go
	reVer := regexp.MustCompile(`^go\d+\.\d+\.(\d+)?$`)
	verStr := runtime.Version()
	m := reVer.FindStringSubmatch(verStr)
	if m == nil {
		return tmp, fmt.Errorf("ebitenmobile: unexpected Go version: %s", verStr)
	}
	toolchainVersion := m[0]
```

Comment (TotallyGamerJet) [reply cmd/ebitenmobile/gomobile.go:168]: Wouldn’t `go/version.Lang(runtime.Version)` give the value you want?

Comment (hajimehoshi) [reply cmd/ebitenmobile/gomobile.go:168]: I didn't know that, thanks!

Comment (hajimehoshi) [reply cmd/ebitenmobile/gomobile.go:168]: https://go.dev/play/p/mJgkLAS_nEw This doesn't work unfortunately.

Comment (hajimehoshi) [reply cmd/ebitenmobile/gomobile.go:168]: https://go.dev/play/p/44G5Md_xRsZ Also `version.Lang` strips the build version, so this is not ideal.

Comment (corfe83) [reply cmd/ebitenmobile/gomobile.go:168]: Good feedback, I've pushed a change as suggested.

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!

File: cmd/ebitenmobile/gomobile.go (hunk 1 above)
Comment (hajimehoshi) [reply cmd/ebitenmobile/gomobile.go]: Hmm, this might not work with gotip: https://go.dev/play/p/V9tVDS4adLm?v=gotip

We need to 1. split the string by a space char and 2. check each chunk whether it maches with a regular expression `^go\d+\.\d+(\.\d+)?`

**Review by hajimehoshi — CHANGES_REQUESTED** [review]
See the comment.

**Comment by corfe83** [comment]
Thanks you're right, this didn't work with tip. I've added the regex, I think it behaves much better now. Also I added *.exe to .gitignore.

File: .gitignore [hunk 2]
```diff
@@ -2,6 +2,7 @@
 *~
 *.aar
 *.apk
+*.exe
```
Comment (hajimehoshi) [inline .gitignore:5]: Why is this needed?

File: cmd/ebitenmobile/gomobile.go [hunk 3]
```diff
@@ -212,6 +214,19 @@ import (
 	return tmp, nil
 }
 
+func getToolchainParameter() string {
+	pattern := regexp.MustCompile("\\bgo\\d+\\.\\d+(\\.\\d+)?")
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:218]: Use ` for a string literal instead of `"`.

File: cmd/ebitenmobile/gomobile.go [hunk 4]
```diff
@@ -212,6 +214,19 @@ import (
...
+	rawVersion := runtime.Version()
+	allIndexes := pattern.FindAllSubmatchIndex([]byte(rawVersion), -1)
+	if len(allIndexes) < 1 {
+		panic("runtime.Version() regex did not capture any results!")
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:222]: Don't panic. Return an error.

File: cmd/ebitenmobile/gomobile.go [hunk 5]
```diff
@@ -212,6 +214,19 @@ import (
...
+	}
+	if len(allIndexes[0]) < 2 {
+		panic("runtime.Version() regex captured a result but didn't return 2 indexes!")
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:225]: Ditto

File: cmd/ebitenmobile/gomobile.go [hunk 6]
```diff
@@ -212,6 +214,19 @@ import (
 	return tmp, nil
 }
 
+func getToolchainParameter() string {
+	pattern := regexp.MustCompile("\\bgo\\d+\\.\\d+(\\.\\d+)?")
+	rawVersion := runtime.Version()
+	allIndexes := pattern.FindAllSubmatchIndex([]byte(rawVersion), -1)
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:220]: Use `FindStringSubmatch`.

File: cmd/ebitenmobile/gomobile.go [hunk 7]
```diff
@@ -212,6 +214,19 @@ import (
 	return tmp, nil
 }
 
+func getToolchainParameter() string {
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:217]: `func toolchainParameter() string`

File: .gitignore (hunk 2 above)
Comment (corfe83) [reply .gitignore:5]: I'm developing on windows, when I run "go build" from "ebiten\cmd\ebitenmobile" directory it produces an exe. I felt we wouldn't want to checkin built binaries to ebiten repo (if we later want to, we could change this). I can remove it if you think this is not appropriate.

**Comment by corfe83** [comment]
Thanks, I've pushed changes to address feedback.

File: .gitignore (hunk 2 above)
Comment (hajimehoshi) [reply .gitignore:5]: This seems a different issue from what we are focusing on now. Let's remove this.

File: cmd/ebitenmobile/gomobile.go [hunk 8]
```diff
@@ -212,6 +218,16 @@ import (
 	return tmp, nil
 }
 
+func toolchainParameter() (string, error) {
+	pattern := regexp.MustCompile(`\bgo\d+\.\d+(\.\d+)?`)
+	rawVersion := runtime.Version()
+	results := pattern.FindStringSubmatch(rawVersion)
+	if len(results ) < 1 {
+		return "", errors.New("runtime.Version() regex did not capture any results!")
+	}
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:227]:
```go
if len(results) == 0 {
    return "", fmt.Errorf("ebitenmobile: unexpected version: %s", rawVersion)
}
```

Also, please do not forget to run gofmt.

File: cmd/ebitenmobile/gomobile.go [hunk 9]
```diff
@@ -212,6 +218,16 @@ import (
 	return tmp, nil
 }
 
+func toolchainParameter() (string, error) {
+	pattern := regexp.MustCompile(`\bgo\d+\.\d+(\.\d+)?`)
+	rawVersion := runtime.Version()
+	results := pattern.FindStringSubmatch(rawVersion)
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:224]: For my convention, `m` is better than `results`. (`m` indicates 'matches')

File: cmd/ebitenmobile/gomobile.go [hunk 10]
```diff
@@ -164,7 +166,11 @@ import (
 	// runtime.Version() is the current executing Go version. For example, this is the version of the toolchain directive in go.mod.
 	// This might differ from the Go command version under the temporary directory.
 	// To avoid the version mismatch, set the toolchain explicitly (#3086).
-	if err := runGo("mod", "edit", "-toolchain="+runtime.Version()); err != nil {
+	toolchainVersion, err := toolchainParameter()
```
Comment (hajimehoshi) [inline cmd/ebitenmobile/gomobile.go:169]: As the variable lifetime is short enough, t is fine.

```go
t, err := toolchainParameter()
```

File: cmd/ebitenmobile/gomobile.go (hunk 8 above)
Comment (corfe83) [reply cmd/ebitenmobile/gomobile.go:226]: Sorry, I hastily ran go fmt and missed the other part of your comment. This is addressed with latest push.

**Review by hajimehoshi — APPROVED** [review]
LGTM, thanks!
